packages/@aws-cdk/aws-ec2/lib/vpc-endpoint-service.ts

```diff
@@ -0,0 +1,87 @@
+import { ArnPrincipal } from '@aws-cdk/aws-iam';
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { CfnVPCEndpointService, CfnVPCEndpointServicePermissions } from './ec2.generated';
+
+/**
+ * A VPC endpoint service.
+ */
+export interface IVpcEndpointService extends IResource {
+}
+
+/**
+ * A VPC endpoint service
+ * @resource AWS::EC2::VPCEndpointService
+ */
+export class VpcEndpointService extends Resource implements IVpcEndpointService {
+
+  /**
+   * One or more network load balancer ARNs to host the service.
+   * @attribute
+   */
+  public readonly networkLoadBalancerArns: string[];
+
+  /**
+   * Whether to require manual acceptance of new connections to the service.
+   * @attribute
+   */
+  public readonly acceptanceRequired: boolean;
+
+  /**
+   * One or more Principal ARNs to allow inbound connections to.
+   * @attribute
+   */
+  public readonly whitelistedPrincipals: ArnPrincipal[];
+
+  private readonly endpointService: CfnVPCEndpointService;
+
+  constructor(scope: Construct, id: string, props: VpcEndpointServiceProps) {
+    super(scope, id);
+
+    if (props.networkLoadBalancerArns === undefined || props.networkLoadBalancerArns.length === 0) {
+      throw new Error("VPC Endpoint Service must have at least one `Network Load Balancer ARN` specified.");
+    }
+
+    this.networkLoadBalancerArns = props.networkLoadBalancerArns;
+    this.acceptanceRequired = props.acceptanceRequired !== undefined ? props.acceptanceRequired : true;
+    this.whitelistedPrincipals = props.whitelistedPrincipals !== undefined ? props.whitelistedPrincipals : [];
+
+    this.endpointService = new CfnVPCEndpointService(this, id, {
+      networkLoadBalancerArns: props.networkLoadBalancerArns,
+      acceptanceRequired: this.acceptanceRequired
+    });
+
+    new CfnVPCEndpointServicePermissions(this, id + "Permissions", {
```

Comment:
If whitelistedPrincipals is empty, is this going to create some weird empty VPCEndpointServicePermission?

Reply:
Yes, however that is allowable CloudFormation syntax. I tested deploying an endpoint service with an empty list of principals and it deployed successfully.

Reply:
I understand that it’s valid CFN, but would it make sense to just not create the resource if whitelistedPrincipals is empty? My fear is that an empty one is created here, then new principals are manually added later, then this resource definition changes and CFN attempts to replace the permissions resource and you’ve just DoS’d yourself.

Reply:
> then new principals are manually added later

Outside of CDK, you mean?

Reply:
I added this in